### PR TITLE
Database registration changes

### PR DESCRIPTION
**Pull request by CDimonaco**
# Description

This PR aims to change the database registration process in sap systems.
We want to ignore database instances that are Secondary in a system replicated scenario.

We accept only database instances that are **not replicated** or **Primary** instances in a replicated scenario.

## How was this tested?

Automated tests

### CODE REVIEW

File: lib/trento/domain/sap_system/sap_system.ex [hunk 1]
```diff
@@ -88,7 +88,18 @@ defmodule Trento.Domain.SapSystem do
...
   end
 
+  def execute(
+        %SapSystem{},
+        %RegisterDatabaseInstance{
+          system_replication: "Secondary",
+          system_replication_status: "ACTIVE"
+        }
+      ),
+      do: {:error, :database_instance_not_primary}
```
Comment (fabriziosestito) [inline lib/trento/domain/sap_system/sap_system.ex:99]: The error should be `:sap_system_not_registered` to indicate that an instance was discovered but no sap system was registered yet. Keep in mind that those domain errors are then bubbling up to the fallback controller, in this case we return a 404.

**Review by fabriziosestito — CHANGES_REQUESTED** [review]

File: lib/trento/domain/sap_system/sap_system.ex (hunk 1 above)
Comment (CDimonaco) [reply lib/trento/domain/sap_system/sap_system.ex:99]: Right, fix coming asap

Comment (CDimonaco) [reply lib/trento/domain/sap_system/sap_system.ex]: done

File: lib/trento/domain/sap_system/sap_system.ex [hunk 2]
```diff
@@ -28,8 +28,9 @@ defmodule Trento.Domain.SapSystem do
 
   That being said, this is the logical order of events in order to register a full system:
 
-  1. A SAP system discovery message with a new database instance is received. At this point, the
-     registration process starts and the database is registered.
+  1. A SAP system discovery message with a new database instance is received.
+     We ignore database with Secondary role in a system replicated scenarios.
```
Comment (arbulu89) [inline lib/trento/domain/sap_system/sap_system.ex:32]: I would avoid using `we` term. `Database instances with Secondary role in a system replication scenario are discarded`

File: lib/trento/domain/sap_system/sap_system.ex [hunk 3]
```diff
@@ -88,7 +89,18 @@ defmodule Trento.Domain.SapSystem do
     embeds_one :application, Application
   end
 
+  def execute(
+        %SapSystem{},
+        %RegisterDatabaseInstance{
+          system_replication: "Secondary",
+          system_replication_status: "ACTIVE"
```
Comment (arbulu89) [inline lib/trento/domain/sap_system/sap_system.ex:96]: I wouldn't honestly check the status field. I would only check what the sap admin configured, which is the role, `Primary/Secondary`. 
Otherwise, we make everything much more complicated, even in the deregistration, as we would need to handle more scenarios.

File: lib/trento/domain/sap_system/sap_system.ex [hunk 4]
```diff
@@ -88,7 +89,18 @@ defmodule Trento.Domain.SapSystem do
     embeds_one :application, Application
   end
 
+  def execute(
+        %SapSystem{},
```
Comment (arbulu89) [inline lib/trento/domain/sap_system/sap_system.ex:93]: Shouldn't we match with `%SapSystem{sap_system_id: nil},` ? otherwise, what happens with sap systems with primary instances registered?

File: lib/trento/domain/sap_system/sap_system.ex [hunk 5]
```diff
@@ -88,7 +89,18 @@ defmodule Trento.Domain.SapSystem do
...
+        %RegisterDatabaseInstance{
+          system_replication: "Secondary",
+          system_replication_status: "ACTIVE"
+        }
+      ),
+      do: {:error, :sap_system_not_registered}
+
   # First time that a Database instance is registered, the SAP System starts its registration process.
+  # We accept the database instance when the system replication is disabled or when enabled, only if the database
```
Comment (arbulu89) [inline lib/trento/domain/sap_system/sap_system.ex:102]: Let's update the `we`

File: lib/trento/application/usecases/clusters/clusters.ex [hunk 6]
```diff
@@ -126,9 +126,10 @@ defmodule Trento.Clusters do
        }) do
     hosts_data =
       Repo.all(
-        from h in HostReadModel,
+        from(h in HostReadModel,
```
Comment (arbulu89) [inline lib/trento/application/usecases/clusters/clusters.ex:129]: why this change? we have a lot of entries without parenthesis

File: test/trento/domain/sap_system/sap_system_test.exs [hunk 7]
```diff
@@ -26,7 +26,138 @@ defmodule Trento.SapSystemTest do
   alias Trento.Domain.SapSystem
 
   describe "SAP System registration" do
-    test "should create an incomplete SAP system aggregate and register a database instance" do
+    test "should fail when a sap system not exists and the database instance has Secondary role with system replication ACTIVE" do
```
Comment (arbulu89) [inline test/trento/domain/sap_system/sap_system_test.exs:29]: `should discard sap system discovery messages from a Secondary role host if the system is not registered`

or at least change `system not exists` by `system which does not exist`

File: test/trento/domain/sap_system/sap_system_test.exs [hunk 8]
```diff
@@ -26,7 +26,138 @@ defmodule Trento.SapSystemTest do
...
+          system_replication_status: "ACTIVE"
+        )
+
+      assert_error(
+        command,
+        {:error, :sap_system_not_registered}
+      )
+    end
+
+    test "should fail when a sap system already exists and the database instance has Secondary role with system replication ACTIVE" do
```
Comment (arbulu89) [inline test/trento/domain/sap_system/sap_system_test.exs:42]: I don't understand this scenario. Once we have a sap system with a primary, other secondary nodes should be added to the instances list

File: test/trento/domain/sap_system/sap_system_test.exs [hunk 9]
```diff
@@ -26,7 +26,138 @@ defmodule Trento.SapSystemTest do
...
+          sid: sid,
+          tenant: tenant,
+          instance_number: instance_number,
+          instance_hostname: instance_hostname,
+          features: features,
+          http_port: http_port,
+          https_port: https_port,
+          start_priority: start_priority,
+          host_id: host_id,
+          system_replication: "Primary",
```
Comment (arbulu89) [inline test/trento/domain/sap_system/sap_system_test.exs:110]: In this case the system replication is enabled, as you have the `Primary` thing.
In a non SR setup, both `system_replication` and `system_replication_status` should be set to nil

**Review by arbulu89 — CHANGES_REQUESTED** [review]
Many comments from my side. I find the logic wrong.
Besides that, I wouldn't use the `status` field, as this is the runtime status of the system replication. With the `system_replication` field, which is what the sap admin sets should be enough

File: lib/trento/application/usecases/clusters/clusters.ex (hunk 6 above)
Comment (CDimonaco) [reply lib/trento/application/usecases/clusters/clusters.ex:129]: Idk, I think the rebase did some magic, here

File: test/trento/domain/sap_system/sap_system_test.exs (hunk 9 above)
Comment (CDimonaco) [reply test/trento/domain/sap_system/sap_system_test.exs]: set accordingly to the test description

File: test/trento/domain/sap_system/sap_system_test.exs (hunk 8 above)
Comment (CDimonaco) [reply test/trento/domain/sap_system/sap_system_test.exs]: yes, removed

File: test/trento/domain/sap_system/sap_system_test.exs (hunk 7 above)
Comment (CDimonaco) [reply test/trento/domain/sap_system/sap_system_test.exs]: fixed

File: lib/trento/domain/sap_system/sap_system.ex (hunk 5 above)
Comment (CDimonaco) [reply lib/trento/domain/sap_system/sap_system.ex]: done

File: lib/trento/domain/sap_system/sap_system.ex (hunk 4 above)
Comment (CDimonaco) [reply lib/trento/domain/sap_system/sap_system.ex]: done man thanks

File: lib/trento/domain/sap_system/sap_system.ex (hunk 3 above)
Comment (CDimonaco) [reply lib/trento/domain/sap_system/sap_system.ex]: rixed

File: lib/trento/domain/sap_system/sap_system.ex (hunk 2 above)
Comment (CDimonaco) [reply lib/trento/domain/sap_system/sap_system.ex]: done

**Comment by CDimonaco** [comment]
Changed, thanks guys, ready to re-review

File: lib/trento/domain/sap_system/sap_system.ex [hunk 10]
```diff
@@ -88,7 +89,17 @@ defmodule Trento.Domain.SapSystem do
...
+  def execute(
+        %SapSystem{sap_system_id: nil},
+        %RegisterDatabaseInstance{
+          system_replication: "Secondary"
+        }
+      ),
+      do: {:error, :sap_system_not_registered}
+
   # First time that a Database instance is registered, the SAP System starts its registration process.
+  # Database instance are accepted when the system replication is disabled or when enabled, only if the database
```
Comment (arbulu89) [inline lib/trento/domain/sap_system/sap_system.ex:101]: `Database instance"s"` add s

File: test/trento/domain/sap_system/sap_system_test.exs [hunk 11]
```diff
@@ -26,7 +26,93 @@ defmodule Trento.SapSystemTest do
   alias Trento.Domain.SapSystem
 
   describe "SAP System registration" do
-    test "should create an incomplete SAP system aggregate and register a database instance" do
+    test "should fail when a sap system not exists and the database instance has Secondary role" do
```
Comment (arbulu89) [inline test/trento/domain/sap_system/sap_system_test.exs:29]: I would add a new test case to see that Secondary nodes are registered once the primary is there.
I think we miss this part, and it would be good to check

File: lib/trento/application/usecases/clusters/clusters.ex (hunk 6 above)
Comment (fabriziosestito) [reply lib/trento/application/usecases/clusters/clusters.ex:129]: please remove the parenthesis, the formatter did something wrong

File: test/trento/domain/sap_system/sap_system_test.exs (hunk 11 above)
Comment (fabriziosestito) [reply test/trento/domain/sap_system/sap_system_test.exs:29]: we already have this test: https://github.com/trento-project/web/blob/9f8e805d08433d0b53a584f7ba3802467840f709/test/trento/domain/sap_system/sap_system_test.exs#L190 

I wouldn't add an extra test since it will be redundant

Comment (arbulu89) [reply test/trento/domain/sap_system/sap_system_test.exs:29]: Alright! It is not explicit, but good enough. I buy your point hehe

**Review by arbulu89 — APPROVED** [review]
Approving. Please fix the test description before merging

File: lib/trento/domain/sap_system/sap_system.ex (hunk 10 above)
Comment (CDimonaco) [reply lib/trento/domain/sap_system/sap_system.ex]: done

File: test/trento/domain/sap_system/sap_system_test.exs (hunk 11 above)
[the remark this answers is not in the document]
Comment (CDimonaco) [reply test/trento/domain/sap_system/sap_system_test.exs]: done

File: lib/trento/application/usecases/clusters/clusters.ex (hunk 6 above)
Comment (CDimonaco) [reply lib/trento/application/usecases/clusters/clusters.ex]: done

**Review by fabriziosestito — APPROVED** [review]
